cli/cook/subcommands/tail.py

```diff
@@ -0,0 +1,176 @@
+        if len(lines) > 1:
+            index_first_delimiter = len(lines[0])
+            partial_line_buffer = partial_line_buffer[:index_first_delimiter]
+            line_buffer = lines[1:] + line_buffer
+
+        # Check if we've read enough lines
+        if check_enough_lines_read(line_buffer, num_lines_to_print):
+            break
+
+        # Check if we've reached the start of the file
+        if check_start_of_file(offset, partial_line_buffer, line_buffer):
+            break
+
+        # Update our offset and length
+        new_offset = max(offset - CHUNK_SIZE, 0)
+        length = offset - new_offset
+        offset = new_offset
+
+
+def tail_follow(file_size, read_fn, follow_sleep_seconds):
+    """Follows the file as it grows, printing new contents"""
+    offset = file_size
+    length = CHUNK_SIZE
+    while True:
+        resp = read_fn(offset=offset, length=length)
+        data = resp['data']
+        num_chars_read = len(data)
+        if num_chars_read > 0:
+            __print(data)
+            offset = offset + num_chars_read
+
+        time.sleep(follow_sleep_seconds)
+
+
+def tail_for_instance(instance, sandbox_dir, path, num_lines_to_print, follow, follow_sleep_seconds):
+    """
+    Tails the contents of the Mesos sandbox path for the given instance. If follow is truthy, it will
+    try and read more data from the file until the user terminates. This assumes files will not shrink.
+    """
+    read = partial(read_file, instance=instance, sandbox_dir=sandbox_dir, path=path)
+    file_size = read()['offset']
+    tail_backwards(file_size, read, num_lines_to_print)
+    if follow:
+        tail_follow(file_size, read, follow_sleep_seconds)
+
+
+def tail(clusters, args):
+    """Tails the contents of the corresponding Mesos sandbox path by job or instance uuid."""
+    uuids = strip_all(args.get('uuid'))
+    paths = strip_all(args.get('path'))
+    lines = args.get('lines')
+    follow = args.get('follow')
+    sleep_interval = args.get('sleep-interval')
+
+    if len(uuids) > 1:
+        # argparse should prevent this, but we'll be defensive anyway
+        raise Exception(f'You can only provide a single uuid.')
+
+    if len(paths) > 1:
+        # argparse should prevent this, but we'll be defensive anyway
+        raise Exception(f'You can only provide a single path.')
+
+    command_fn = partial(tail_for_instance, path=paths[0], num_lines_to_print=lines,
+                         follow=follow, follow_sleep_seconds=sleep_interval)
+    query_unique_and_run(clusters, uuids[0], command_fn)
+
+
+def register(add_parser, add_defaults):
+    """Adds this sub-command's parser and returns the action function"""
+    parser = add_parser('tail', help='output the last part of files in the Mesos sandbox by job or instance uuid')
+    parser.add_argument('--lines', '-n', help=f'output the last NUM lines (default {DEFAULT_NUM_LINES})',
+                        metavar='NUM', type=check_positive)
```

Comment:
Sorry, I think I mixed up `NUM` here with `N` below. I meant we should test what happens when number-of-lines is set to zero.

Reply:
👍 will do

Reply:
Done